Plot feature loadings function

ElySeraidarian opened the pull request:
According to #92, we need a way to plot feature loadings after performing a reduction method by several ways. Here is a first version with PCA that plots feature loadings in 4 different ways : Tree, Barplot, Screeplot & Heatmap. This will be available for LDA & NMF as soon as it is possible to retrieve loadings matrix from reducedDim slot. There are parameters that can be added so that we can talk about it (for example the number of features plotted). The parameter method is probably doomed to disappear as we will be able to recognize which method has been used in the reducedDim slot. There are examples for NMF & LDA at the end for now without using the function. This will need examples in OMA and strong documentation in order to tell people how to use this function (agglomerate tree if plotting the tree, only plotting few features, ...). It is currently possible to give either loadings matrix or TSE object to plot.

I will add tests & generate documentation also later.

Commit:
Adding confidence level parameter

Commit:
Small verbose changes

Commit:
Documentation changes

antagomir committed:
Merge branch 'devel' into devel

Commit:
Fixing parameter name

Commit:
Merge remote-tracking branch 'origin/devel' into devel

Commit:
Test build

Commit:
Creating function's body

Commit:
Name changes

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

Commit:
Function progresses

Commit:
Complete example

Commit:
Bumped version & update news

Commit:
Fixed boolean statement

Commit:
Real fix boolean statement

Commit:
Started function core & encapsulation

Commit:
Fix if statements

TuomasBorman committed:
Update NEWS

Commit:
Dealing with issues from other reduction methods

Commit:
Merge remote-tracking branch 'origin/devel' into plotFeatureLoadings

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

Commit:
Finished function for example use

Commit:
Added example & removed prints statements

Commit:
Progress function & working examples

ElySeraidarian committed:
Merge branch 'microbiome:devel' into devel

Commit:
All working examples + added heatmap (format to define later)

Commit:
merging

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

Commit:
Merge remote-tracking branch 'origin/devel' into plotFeatureLoadings

Commit:
Added PCA example & comments

Commit:
Added support for no tree use

Commit:
Function V1.0 with working examples

antagomir committed:
Merge branch 'devel' into plotFeatureLoadings

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' Plot tree with feature loadings after performing a reducedDim (e.g. PCA, LDA, NMF)
```
Remove " (e.g. PCA, LDA, NMF)" -> that cannot be on the title line. This can be moved to line 4

antagomir replied:
This method works also without tree -> remove "tree"?

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' Plot tree with feature loadings after performing a reducedDim (e.g. PCA, LDA, NMF)
+#'
+#' Based on the \code{\link[diffTop:plotASVcircular]{plotASVcircular}} method using phyloseq 
```
"Based on" -> "Inspired by"

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' Based on the \code{\link[diffTop:plotASVcircular]{plotASVcircular}} method using phyloseq 
+#' and has been converted to use TreeSummarizedExperiment objects.
+#'
+#' @param object a
```
"x" would be more standard here (instead of object)

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#'   \code{\link[TreeSummarizedExperiment:TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'   object.
+#' 
+#' @param method One dimensionality reduction method of \code{c("PCA", "LDA", "NMF")}
```
Do we need this if the method should also support general numeric loading matrix?

At least this should be optional. Ideally, the method is agnostic to which method generated the loadings.

TuomasBorman replied:
`method` could be `dimred`. You can check plotRDA which plots the principal coordinates

ElySeraidarian replied:
For now it doesn't recognize the reduction method used but it will probably be removed once it is done

antagomir replied:
Ok - can we remove this and resolve the conversation, or is this needed?

TuomasBorman replied:
One problem is that in practice the ordination result can be named abitrarily (it can be "pca" or "pca_something"). There should be a way to determine the fetched ordination from reducedDim.

The result should be calculated with runPCA or addLDA, so the result is standardized. If the loadings are not in certain place, you can just give error.

See plotRDA

antagomir replied:
Yes. 

And loadings are always a numeric matrix. If we can support arbitrary numeric matrices in addition to TreeSE slot here then that would be worth having I think.

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#'   indicating which method will be used.
+#'   (default: \code{method = "PCA"})
+#'  
+#' @param rank a single character defining a taxonomic rank. Must be a value of
```
Btw is it possible to show tree and PC components with higher resolution than the colored rank info?

ElySeraidarian replied:
I think this should be possible

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' @param rank a single character defining a taxonomic rank. Must be a value of
+#'   \code{taxonomyRanks()} function.
+#'   
+#' @param plot One way to plot feature loadings of \code{c("Heatmap", "Barplot", "Screeplot", "Tree")} 
```
I suggest to use here argument name "method"

TuomasBorman replied:
`layout` is use in other functions

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' @return 
+#' A \code{ggplot2} object. A circular plot annotated with TreeSummarizedExperiment object.
+#'
+#' @name plotFeatureLoadings
```
I think the @name tag is not needed (?). 

The function name could be shorter, just plotLoadings?

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' library(scater)
+#' data("GlobalPatterns", package = "mia")
+#' tse <- GlobalPatterns
+#' tse <- logNormCounts(tse)
```
this should use instead the transformAssay function

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' data("GlobalPatterns", package = "mia")
+#' tse <- GlobalPatterns
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
```
Any chance to keep lower prevalence threshold? 0.99 is very high.

ElySeraidarian replied:
Yes indeed, it was for my examples to keep very few features

antagomir replied:
How about just aggregating to even higher level? Then you will have less features without need to add prevalence filtering.

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' tse <- GlobalPatterns
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
```
Is it possible to have support for more than 2 PC components in the plotting method?

ElySeraidarian replied:
Yes I can do it

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
+#' plotFeatureLoadings(tse, method= "PCA", plot = "Tree")
```
I would use lowercase versions for the method names: "tree", "barplot" etc

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' 
+#' 
+#' # Plotting without tree as a heatmap
+#' library(mia)
+#' library(ggtree)
+#' library(scater)
+#' data("GlobalPatterns", package = "mia")
+#' tse <- GlobalPatterns
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
```
No need to repeat these steps a second time in the example

antagomir replied:
(same comment below)

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
+#' plotFeatureLoadings(loadings_matrix, method= "PCA", plot = "Screeplot")
```
How about p + coord_flip() as default for better readability of the taxa names?

ElySeraidarian replied:
I tried but then it's just a barplot? I don't know how I could increase readability otherwise

antagomir replied:
Hmm the difference to barplot is that the bars are sorted here?

ElySeraidarian replied:
They should be sorted in both now

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
+#' plotFeatureLoadings(loadings_matrix, method= "PCA", plot = "Barplot")
```
Barplot could be sorted from the smallest to largest value and the names do not have to match between the two PC components.

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+#' tse <- logNormCounts(tse)
+#' tse <- agglomerateByPrevalence(tse, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' tse <- runPCA(tse, name = "PCA", ncomponents = 2)
+#' plotFeatureLoadings(loadings_matrix, method= "PCA", plot = "Heatmap")
```
Could we have just one color legend common to all PC components?

Either a single heatmap, or if heatmaps are separate (per component) then it might work better to sort the values from the smallest to the largest within each PC component.

The color order should be reverted (blue negative, red positive)

antagomir replied:
Title should be PC1, PC2 etc. (not "Feature loadings plot")

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+  }
+  
+  if (is.matrix(object) && plot == "Tree") {
+    stop("Tree cannot be plotted for a non-TSE object.", call. = FALSE)
```
Write positively: "TreeSummarizedExperiment object is required for the tree plotting".

antagomir commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+  
+  
+}
+##############################################
```
Are the lines from 257 onwards necessary?

antagomir reviewed (changes requested):
Great! See the comments.

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+
+
+#' @rdname plotFeatureLoadings
+#' @export 
+setMethod("plotFeatureLoadings", signature = c(object = "TreeSummarizedExperiment"),
+          function(object,
```
General comment:

We have agreed to develop the methods strictly following the Bioconductor guidelines. Run BiocCheck::BiocCheck().

The indentation should always be multiplication of 4 spaces. Also the line should have max 80 characters.

Important thing is to add explanations as comments what the code does. This makes the maintenance lots of easier when we know what was the purpose.

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+              rownames(df2) <- phylo$tip.label
+              
+              
+              p <- gheatmap(
```
Is this gheatmap new dependency?

ElySeraidarian replied:
gheatmap is from ggtree package that is used in others miaViz functions (e.g. plotTree)

TuomasBorman replied:
Thanks, then it is ok

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+            
+            .check_parameters(object,
+                              method = method, plot = plot, ...)   
+            
+            if (method == "PCA") {
+              loadings_matrix <- attr(reducedDim(tse, "PCA"), "rotation")
+              df <- data.frame(loadings_matrix)
+              df[["Feature"]] <- rownames(loadings_matrix)
+              df[["Feature"]] <- factor(df[["Feature"]], levels = df[["Feature"]])
+              df <- df[ order(abs(df[["PC1"]]), decreasing = TRUE)[1:10], ]
+              df <- df[ order(df[["PC1"]]), ]
+              df <- df[ order(abs(df[["PC2"]]), decreasing = TRUE)[1:10], ]
+              df <- df[ order(df[["PC2"]]), ]
+            } else {
+              stop("These methods are yet to be implemented", call. = FALSE)
+            }
+            
+            if (plot == "Tree") {
+              
+              phylo <- rowTree(object)
+              circ <- ggtree(phylo, layout = "circular")
+              df <- rowData(object)
+              color <- randomcoloR::distinctColorPalette(
+                length(
+                  unique(
```
Try to restructure the code. The "main" function should be simple and call internal functions. Here:

1. You could first call `.get_loadings_plot_data` that prepares data.frame from the data
2. The you could call `.loadings_plotter` That creates the plot.

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+            
+            if (plot == "Tree") {
+              
+              phylo <- rowTree(object)
```
What if rowTree() is NULL?

ElySeraidarian replied:
this is checked in .check_parameters before going there

TuomasBorman replied:
OK, good!

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+                  df2, (all_of(i))
+                )
+                
+                p <- gheatmap(
```
INstead of using new dependency, you could check https://github.com/microbiome/miaViz/blob/dde325321015e657ff2cf904c9feb8387c9eb2f4/R/plotPrevalence.R#L576

TuomasBorman replied:
For tree plotting, we have this function https://github.com/microbiome/miaViz/blob/dde325321015e657ff2cf904c9feb8387c9eb2f4/R/plotTree.R#L1031 Try to check if you can use that code.

For maintenance, it is always easier if we can use and improve the existing code rather than having multiple implementations on the same thing (tree plotting might not be relevant here though, just a general note. However, for heatmap the existing code might work)

TuomasBorman commented on R/plotFeatureLoadings.R:
```diff
@@ -0,0 +1,404 @@
+                  legend_title = expression(beta[k]))
+              }
+              
+              p
+              
+            } else {
+              .plot_pca_feature_loadings(df, plot)
+            }
```
It is hard to see when the function ends. I suggest to save the plot to variable and always end the function with return()

Commit:
Review changes

ElySeraidarian commented:
Made some changes, I'll add a parameter to increase the number of principal components and maybe one optional parameter to select the name used for the matrix in reducedDim slot if it has been changed by the user? So that I'll be able to recognize the method used

Commit:
Added support for more components & different naming

ElySeraidarian commented:
I have added support for more PC components (I have not added a parameter, it plots the number of components given by the matrix but maybe I should add if user wants to plot less?).  Also added support if name is changed in reducedDim slot.

antagomir committed:
Merge branch 'devel' into plotFeatureLoadings

antagomir commented:
Yes, it is necessary to give the number of components. Argument name "ncomponents" has been used elsewhere. 

By default it could be ncomponents=5?

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+    # Checking if dimred is correct
+    if ( !(dimred %in% c("PCA", "LDA", "NMF")) ) {
+        stop("'dimred' must be one of c('PCA', 'LDA', 'NMF').", call. = FALSE)
+    }
```
Now user cannot plot reducedDim(tse, "My_PCA")

--> Check this https://github.com/microbiome/miaViz/blob/c1c277a36b8f8ec195cd24544c4fc889ec5dc751/R/plotCCA.R#L313

Update also descrption

TuomasBorman replied:
Also check that teh value is character scalar .is_a_string

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+        stop("'dimred' must be one of c('PCA', 'LDA', 'NMF').", call. = FALSE)
+    }
+    # Checking if layout is correct
+    if ( !(layout %in% c("screeplot", "barplot", "tree", "heatmap")) ) {
```
This fails if user inouts a vector. --> .is_a_string()

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+        stop("TreeSummarizedExperiment object is required for the tree plotting.", call. = FALSE)
+    }
+    # Making sure the tree is not null
+    if (is(x, "TreeSummarizedExperiment")) {
+        if ( is.null(rowTree(x)) && layout == "tree") {
+            stop ("Tree is null.", call. = FALSE)
+        }
+    }
```
Try to avoid nested ifs. Moreover, I think this does not catch if the object is SummariedExperiment.

```
if( layout == "tree" && !(is(x, "TreeSummarizedExperiment") && !is.null(rowTree(x, tree.name)) ){

}
```

TuomasBorman replied:
Moreover, user should be able to specify tree.name (the name of rowTree) --> check other methods

ElySeraidarian replied:
This is because user can decide to give only the loadings matrix in this case it will return error because there is no rowTree in a matrix

ElySeraidarian replied:
Nevermind seems to work, I believed it would go through all the checks in if statement but it stops if first one is incorrect

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+
+.get_loadings_plot_data <- function(x, n, ncomponents) {
+    L <- list()
+    for (i in 1:ncomponents) {
+        df <- data.frame(x)
+        df[["Feature"]] <- rownames(x)
+        # Ordering loadings
+        df <- df[ order(abs(df[[paste("PC",as.character(i),sep="")]])[1:n], decreasing = TRUE), ]
```
I believe you can do this without loop. (Loops should be avoided in R if possible, they are slow. Also good practice to avoid loops)

Moreover, you are handling the data with column names. However, it is not necessary since you know the index of columns. It is preferred to use indices since that is more general approach (column names can differ; they are not always "PC*")

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+}
+
+.loadings_tree_plotter <- function(x, loadings_matrix, ncomponents) {
+    phylo <- rowTree(x)
```
User should be able to specify the tree. --> tree.name parameter

TuomasBorman replied:
Default value "phylo". Moreover, then you should subset the data based on the tree (TreeSE can have multiple trees, and tree might be missing some rows).

tree <- rowTree(tse, tree.name)
ind <- rowLinks(tse)[["whichTree"]] == tree.name
if( any(ind) ){
warning("Data was subsettd")
}
tse_sub <- tse[ind, ]

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,341 @@
+        )
+      )
+    )
+    df <- data.frame(Class = df$Phylum)
+    rownames(df) <- phylo$tip.label
+    
+    df2 <- data.frame(abs(loadings_matrix))
+    rownames(df2) <- phylo$tip.label
+    
+    
+    p <- gheatmap(
+        p = circ,
+        data = df,
+        offset = -.1,
+        width = .1,
+        colnames_angle = 95,
+        colnames_offset_y = .5,
+        font.size = 4,
+        color = "black") +
+        ggplot2::scale_fill_manual(
+            values = color,
+            name = "Class"
+        )
+    
+        for(i in 1:ncomponents){
+            if(i == 1){
+                p <- p +
+                ggnewscale::new_scale_fill()
+            }
+            df3 <- dplyr::select(
+                df2, (all_of(i))
+            )
+        
+            p <- gheatmap(
```
Add comments

Commit:
Added tree_name, removed loop & tests changes

Commit:
Merge branch 'plotFeatureLoadings' of github.com:ElySeraidarian/miaViz into plotFeatureLoadings

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+#'   \code{\link[TreeSummarizedExperiment:TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'   x.
+#' 
+#' @param dimred One dimensionality reduction method of reducedDimNames
+#'   indicating which method will be used.
+#'   (default: \code{dimred = "PCA"})
```
Put this parameter descriptions in this format.

Class. Definition. Default value.

-->

dimred \code{Character scalar}. DImension reduction from reducedDimNames to be plotted. (Default: \code{"PCA"])

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+    if ( !(is.numeric(ncomponents) && ncomponents > 0) ) {
+        stop("'ncomponents' must be a positive number.", call. = FALSE)
+    }
+
```
R functions always return something. I think it is good practice to specify the returned value even though the value is not used. it could be just return(NULL)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+    if( layout == "tree" && !(is(x, "TreeSummarizedExperiment") && !is.null(rowTree(x, tree_name)) )) {
+        stop ("Tree is null.", call. = FALSE)
+    }
+    #Checking if n is a positive number
```
Add space after #

TuomasBorman replied:
Check all the comments and add spaces

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+
+}
+
+.check_components <- function(x, ncomponents) {
+    #Checking if there are enough components in the matrix
+    if (ncomponents > length(colnames(x))) {
+        stop("'ncomponents' must be lower or equal than number of components.", call. = FALSE)  
+    }
+}
```
length(colnames()) -> ncol()

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+                labs(title=paste("PC",i,sep=""))
+        }
+
+        return(p)
```
I think it is more clear that the function returns values in the end of the function. You could move return(p) to end , and remove them from inside if statements

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+#' @author
+#' Ely Seraidarian
+#' Contact: \url{microbiome.github.io}
+#' @examples
```
Add line break above examples

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+#' Ely Seraidarian
+#' Contact: \url{microbiome.github.io}
+#' @examples
+#' 
```
remove extra line

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+                        tree_name = tree_name,
+                        ...)  
+
+        loadings_matrix <- attr(reducedDim(x, dimred), "rotation")
```
This works now with PCA. How about with RDA (mia::getRDA)

ElySeraidarian replied:
This does not work for now, but @thpralas will create a wrapper for every reducedDim methods and store in feature_loadings instead of rotation so it will work for every method. I tried and currently RDA loadings matrix is not stored in rotation slot so it will not work here

TuomasBorman replied:
Wrappers might not be optimal solutions since it requires more maintenance and it might be that users still use scater::runPCA instead of our wrapper. It might be then quite confusing for user why the function does not work even though user used the "correct" function.

Instead, you could create an internal function

```
loading_names <- c("rotation", "loadings", "some_other_name")
attr_names <- names(attributes(reduced_dim))
attr_name <- attr_names[ attr_names %in% loading_names ]
if( length(attr_name) != 1 )){
    stop("Loadings cannot be found..")
}
loadings <- attr(tse, attr_name)
```

TuomasBorman replied:
The code does not necessarily work but gives the idea

ElySeraidarian replied:
I see but for example for RDA there is no attribute where it is stored, so you can't even retrieve the matrix that easily

TuomasBorman replied:
Sorry, last week I was busy with OMA, so I had not time to check this.

From RDA, you can get "constrained weighted orthonormal species scores" which are analogous to loadings of PCA. See for instance this https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/cca.object

When used getRDA, (getCCA) or add* functions, these are stored into "rotation" attribute of returned matrix.

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+                        layout = layout,
+                        n = n,
+                        ncomponents = ncomponents,
+                        tree_name = tree_name,
```
tree_name --> tree.name

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+#' @param ncomponents A numeric specifying the number of components.
+#'   (default: \code{ncomponents = 5})
+#' 
+#' @param tree_name a single \code{character} value specifying a rowTree/colTree from
```
This cannot specify colTree

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+        loadings_matrix <- attr(reducedDim(x, dimred), "rotation")
+        #Checking if there are enough components in the matrix
+        .check_components(loadings_matrix, ncomponents)
+        # Ordering loadings and adding factor to keep the order
+        L <- .get_loadings_plot_data(loadings_matrix, n, ncomponents)
```
THis is not used if layout == "tree". You can move it inside` else`

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+            layout = "heatmap",
+            n = 10,
+            ncomponents = 5,
+            tree_name = "phylo",
```
tree_name is not used in this function?

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+  # Transform into a dataframe
+  df <- data.frame(x)
+  # Keep only the number of components needed
+  names <- colnames(df)[1:ncomponents]
+  df <- dplyr::select(df, all_of(names))
```
You could just perhaps:

df <- df[, seq_len(ncomponents), drop = FALSE]

ElySeraidarian replied:
I tried and somehow it wasn't working so I went back to previous version (maybe I made a mistake somewhere)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,374 @@
+
+.get_loadings_plot_data <- function(x, n, ncomponents) {
+  # Transform into a dataframe
+  df <- data.frame(x)
```
as.data.frame()

TuomasBorman replied:
To save memory, it might be better to store the data back to x

TuomasBorman reviewed (changes requested):
Couple comments

Commit:
Added tests & minor changes

ElySeraidarian commented:
Hadn't seen comments I'll check this

Commit:
Review changes

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

Commit:
Small  plotting changes

Commit:
Changed colors for tree plot

Commit:
Merge remote-tracking branch 'upstream/devel' into plotFeatureLoadings

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+#'   x.
+#' 
+#' @param dimred \code{Character scalar}. Dimension reduction from reducedDimNames to be plotted. 
+#'   (default: \code{dimred = "PCA"})
```
(Default: \code{"PCA"}) is simpler. Other functions follow that style

TuomasBorman replied:
Modify also other parameter descriptions

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+    # Subset data based on the tree
+    ind <- rowLinks(x)[["whichTree"]] == tree.name
+    if( any(ind) ){
+      warning("Data was subsetted")
```
warning("message here", call. = FALSE)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+        }
+    }
+    
+    else if (layout == "screeplot") {
+        # Transform into a dataframe
+        df <- data.frame(PC = L[[1]][[1]], Feature = L[[1]][["Feature"]]) 
+        # Plot first component
+        p <- ggplot(df, aes(x = .data[["PC"]], .data[["Feature"]])) +
+            geom_bar(stat="identity") +
+            theme(axis.title.x = element_blank(), 
```
This is not scree plot. Scree plot includes eigenvalues

TuomasBorman replied:
As this is same as bar plot, this can be removed

TuomasBorman commented:
For heatmap, "shadow text" might be better

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+
+#' @rdname plotLoadings
+#' @export 
+setMethod("plotLoadings", signature = c(x = "TreeSummarizedExperiment"),
```
We could enable the function for SingleCellExperiment objects also but then missing rowTree slot should be taken into acount (SCE objects do not have trees)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+  L <- lapply(1:ncomponents, .process_component, df = x, n = n)
+  return(L)
+}
+
```
Remove extra line break

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+    circ <- ggtree::ggtree(phylo, layout = "circular")
+    df <- rowData(x)
+    # Transform into a dataframe
+    df <- data.frame(Class = df$Phylum)
```
User should be able to choose the rank (or either no rank at all, i.e., rownames)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+    # Transform into a dataframe
+    df <- data.frame(Class = df$Phylum)
+    # Match labels
+    rownames(df) <- phylo$tip.label
```
This can cause serious problems. That is because rows

1. do not have to match with tips
2. The order might not be the same in rows and nodes of tree

You can get the linkage between rows and nodes of tree from rowLinks()

TuomasBorman replied:
The point 1 can cause problems even after rowLinks() since your method is assuming that rows are in tips of tree?

TuomasBorman replied:
Although, you could check whether the rows are in tips, and if they are not, then you could give informative error

TuomasBorman replied:
Then if the tree is not agglomerated in prior to this function, you could prune it with mia::.prune_tree

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+  x <- as.data.frame(x)
+  # Keep only the number of components needed
+  names <- colnames(x)[1:ncomponents]
+  df <- dplyr::select(x, all_of(names))
```
Instead of dplyr::select(), use importFrom above the function

@importFrom dplyr select

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+    }
+    x <- x[ind, ]
+    # Store plot tree
+    circ <- ggtree::ggtree(phylo, layout = "circular")
```
Here also. Add

#' @importFrom ggtree ggtree

above

.loadings_tree_plotter

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,362 @@
+            xlim(-1,1) +
+            labs(title=names[1]) +
+            coord_flip()
+        # Plot others components
+        for (i in 2:length(L)) {
+            df <- data.frame(PC =L[[i]][[i]],
+                Feature = L[[i]][["Feature"]])
+            p <- p + ggplot(df, aes(x = .data[["PC"]], .data[["Feature"]])) +
+                geom_bar(stat="identity") +
+                theme(axis.title.x = element_blank(),
+                    axis.text.x = element_text(angle = 90,
+                        vjust = 0.5, hjust=1, size=6)) +
+                xlim(-1,1) +
+                labs(title=names[i]) +
+                coord_flip()
+        }
```
Instead of looping like this, you could use facets which is more neat solution (then all the data must be in same data.frame)

TuomasBorman replied:
```
# Prepare the data in correct format, this might not take into account
# the correct order and features, but it gives the idea
df <- lapply(seq_len(length(L)), function(i){
    L[[i]][, i, drop = FALSE]
})
df <- do.call(cbind, df) 
df[["feat"]] <- rownames(df)

# To long format
df <- reshape(
    df, varying =colnames(df)[ !colnames(df) %in% c("feat") ],
    v.names = "value", 
    timevar = "PC",
    times = colnames(df)[ !colnames(df) %in% c("feat") ],
    direction = "long")

ggplot(df, aes(x = .data[["value"]], y = .data[["feat"]])) +
    geom_bar(stat="identity") +
    facet_wrap(~ PC, scales = "free")

```

TuomasBorman replied:
Check also other plot types. I think the data.frame list might not be needed. Just use facets instead of looping over list of data.frames

ElySeraidarian replied:
This is indeed a nice solution but after trying for hours, haven't been able to make axes label change for every component like it used to be. Also ordering seems to be working only on first component for some reason. 
![image](https://github.com/user-attachments/assets/a21f8658-2f1d-40d7-b436-97e51088e1ba)
Here is the best I have been able to do but I guess it looks worse

antagomir replied:
I think you would need to define every (sub)figure separately in order to have a full control.

TuomasBorman replied:
How about this?

```
library(tidytext)
ggplot(df, aes(x = Value, y = reorder_within(Feature, Value, PC))) +
    geom_bar(stat = "identity") +
    scale_y_reordered() +
    facet_wrap(~ PC, scales = "free") +
    theme_minimal() +
    labs(x = "Value", y = "Feature")
```

TuomasBorman replied:
This works for heatmap
```

library(tidytext)
library(shadowtext)
ggplot(df, aes(x = .data[["PC"]], y = reorder_within(Feature, Value, PC), label = round(.data[["Value"]], 2)))  +
    geom_point(aes(fill = .data[["Value"]]), size = 15-0.4*n, shape = 22) +
    scale_fill_gradient2(
        limits = c(-1,1), low = "darkslateblue", mid = "white", high = "darkred") +
    scale_y_reordered() +
    geom_shadowtext(size = 4-0.075*n) +
    facet_wrap(~ PC, scales = "free")
```

antagomir committed:
Merge branch 'devel' into plotFeatureLoadings

Commit:
Removed loops for plottings, added SCE support & class parameter

Commit:
Merge branch 'plotFeatureLoadings' of github.com:ElySeraidarian/miaViz into plotFeatureLoadings

Commit:
Fixed issues & matching labels for tree

ElySeraidarian commented:
Every change should have been done according to your comments, the values in tree layout is now accurately matching with labels (getting same results with heatmap and barplot). I'll focus on making it operational for every reduction method (detecting which method has been performed) with what @thpralas has done ( and maybe find a solution if user is not aware of these wrappers)

antagomir commented:
Just for clarity, could you "resolve" all comments that have been addressed?

antagomir committed:
Merge branch 'devel' into plotFeatureLoadings

antagomir commented:
Solution for non-wrapper users would be to support plain numeric matrices (without tree) in addition to (Tree)SE objects. If users want to include a tree then I think it is better to require that they create a TreeSE object.

antagomir commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,385 @@
+#' Plot feature loadings after performing a reducedDim 
+#'
+#' Inspired by the \code{\link[diffTop:plotASVcircular]{plotASVcircular}} method using phyloseq 
+#' and has been converted to use TreeSummarizedExperiment objects.
```
I would here include one sentence description of what this function does.

And then the current lines 3-4 could go to @details section.

antagomir commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,385 @@
+#' Plot feature loadings after performing a reducedDim 
```
Change this to: "Plot feature loadings for TreeSummarizedExperiment objects."

Then later in @details explain that the TreeSE object is expected to have content in reducedDim slot.

However TreeSE inherits SingleCellExperiment, and SingleCellExperiments have reducedDim slot (but no tree). Could we make this more generic and add support for SCE objects?

antagomir commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,385 @@
+#' Inspired by the \code{\link[diffTop:plotASVcircular]{plotASVcircular}} method using phyloseq 
+#' and has been converted to use TreeSummarizedExperiment objects.
+#'
+#' @param x a
```
TreeSummarizedExperiment or SingleCellExperiment? (the latter does not have a tree and it is more generic, TreeSE is a SCE).

ElySeraidarian replied:
Yes, should have updated description but I added SCE support in last commits.

antagomir commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,385 @@
+#'   (default: \code{rownames})
+#'     
+#' 
+#' @details
```
You can also say here something about the limitations related to feature and sample set sizes etc.

Commit:
Description changes

Commit:
Support for other methods

Commit:
Merge branch 'plotFeatureLoadings' of github.com:ElySeraidarian/miaViz into plotFeatureLoadings

Commit:
Correct example & added length-1 reducedDim slot autodetection

ElySeraidarian commented:
I have been able to make it work for LDA method with the wrapper 
``` 
library(mia)
library(ggtree)
library(scater)
data("GlobalPatterns", package = "mia")
tse <- GlobalPatterns
tse <- transformAssay(tse, method = "clr", pseudocount = 1)
tse <- agglomerateByPrevalence(tse, rank="Phylum", update.tree = TRUE)
tse <- addLDA(tse)
plotLoadings(tse,  ncomponents = 2)
``` 
However I have not been able to autodetect method if reducedDim slot contains several methods, then user will have to specify name using dimred = "method_name" and I don't believe this is possible either?

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+            class = rownames(x),
+            ...) {
+      
+        if (length(reducedDimNames(x)) == 1) {
+            dimred <- reducedDimNames(x)
+        }
```
See other functions that plot reducedDim slot. I think, default value of dimred should not be determined and this line can be removed.

1. The functions should be similar
2. The functions should be transparent. In practice, this does differently than user has specified

ElySeraidarian replied:
Maybe I could still keep it so that user do not have to specify dimred when only one reduction method has been performed (removed default value though)

TuomasBorman replied:
I am not fan of this solution since it is not transparent. we should allow user to get error if user causes it. If we do not do that it is possible source of error. at least it should be

dimred =reducrddimnanmes(tse)[[1]]

In the function definition

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+        attr_names <- names(attributes(reducedDim(x, dimred)))
+        attr_name <- attr_names[ attr_names %in% loading_names ]
+        if( length(attr_name) != 1 ) {
+            stop("Loadings cannot be found..")
```
call. = FALSE

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+    # Checking if class is correct
+    if ( is(x, "TreeSummarizedExperiment") && !(any(class %in% taxonomyRanks(x)) || identical(class, rownames(x)))) {
+        stop("'class' must be one of taxonomyRanks", call. = FALSE)
+    }
```
We use "rank" in other methods. Should be the same

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+#'   TreeSummarizedExperiment object. (By default: \code{"phylo"})
+#'   
+#' @param class A single \code{character} value specifying a rank from taxonomyRanks
+#'   (default: \code{rownames})
```
The parameter name could be "rank". Moreover, the default value could be NULL --> which means that no rank from rowData

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+#' @param layout One way to plot feature loadings of \code{c("heatmap", "barplot", "tree")} 
+#'   (default: \code{"tree"} for TSE, \code{"heatmap"} for others)
+#' 
+#' @param n A numeric specifying the number of features to be plotted.
```
Check all the parameter descriptions. They should follow the same converntion that in the other functions

Class. Description. Default

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+                    plot.title = element_text(size = 18, hjust=0.5))
+        }
+    return(p)
+}
```
Legend title is now "Class" but it could be the specified rank

Moreover, usually titles are not added to article figures, so it might be better to not add the title by default (User can easily add it afterwards)

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+setMethod("plotLoadings", signature = c(x = "TreeSummarizedExperiment"),
+    function(x,
+            dimred = "PCA",
+            layout = "tree",
```
I would keep the default layout same for TreeSE

TuomasBorman replied:
Also while tree layout looks nice, it is rather complex.

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+      name <- colnames(pc_df)
+      new_df[rownames(pc_df), name] <- pc_df[, 1]
+    }
+    new_df[["Feature"]] <- names
```
Instead of construting the single df in here, I think it might be much neater that the input is already one single data.frame. That helps also the maintenance.

antagomir commented:
COuld you use `reducedDimNames(tse)` to detect?

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,416 @@
+.check_parameters <- function(x, dimred, layout, n, ncomponents, tree.name, class, ...) {
+    # Check tree.name
+    if( is(x, "TreeSummarizedExperiment") && !(tree.name %in% rowTreeNames(x)  && .is_a_string(tree.name))){
+        stop("'tree.name' must be a single character value specifying a colTree.", call. = FALSE)
```
rowTree

TuomasBorman replied:
Also this is failing while it should not

```
data("GlobalPatterns", package = "mia")
tse <- GlobalPatterns
tse <- transformAssay(tse, method = "clr", pseudocount = 1)
tse <- agglomerateByPrevalence(tse, rank="Phylum", update.tree = TRUE)
tse <- runPCA(tse, ncomponents = 5, assay.type = "clr")
plotLoadings(tse, layout = "barplot")
```

TuomasBorman replied:
Not all TreeSE have tree

ElySeraidarian replied:
> Also this is failing while it should not
> 
> ```
> data("GlobalPatterns", package = "mia")
> tse <- GlobalPatterns
> tse <- transformAssay(tse, method = "clr", pseudocount = 1)
> tse <- agglomerateByPrevalence(tse, rank="Phylum", update.tree = TRUE)
> tse <- runPCA(tse, ncomponents = 5, assay.type = "clr")
> plotLoadings(tse, layout = "barplot")
> ```

I don't see why this would not work? It runs fine for me

TuomasBorman replied:
Forgot to add rowTree(tse) <- NULL

data("GlobalPatterns", package = "mia")
tse <- GlobalPatterns
rowTree(tse) <- NULL
tse <- transformAssay(tse, method = "clr", pseudocount = 1)
tse <- agglomerateByPrevalence(tse, rank="Phylum", update.tree = TRUE)
tse <- runPCA(tse, ncomponents = 5, assay.type = "clr")
plotLoadings(tse, layout = "barplot")

TuomasBorman commented:
For default layout, I would keep barplot as it is the most simple and fastest to look

TuomasBorman commented:
> I have been able to make it work for LDA method with the wrapper
> 
> ```
> library(mia)
> library(ggtree)
> library(scater)
> data("GlobalPatterns", package = "mia")
> tse <- GlobalPatterns
> tse <- transformAssay(tse, method = "clr", pseudocount = 1)
> tse <- agglomerateByPrevalence(tse, rank="Phylum", update.tree = TRUE)
> tse <- addLDA(tse)
> plotLoadings(tse,  ncomponents = 2)
> ```
> 
> However I have not been able to autodetect method if reducedDim slot contains several methods, then user will have to specify name using dimred = "method_name" and I don't believe this is possible either?

That is ok in my opinion. Other functions also assume that user specifies the reducedDim to plot.

For instance:

plotReducedDim(tse, dimred = "PCA") --> you have to always specify dimred in this function

TuomasBorman commented:
I am also wondering if current heatmap layout takes full advantage of heatmaps. Heatmaps are create way to visualize large batch of data. However, in this layout it rather makes the plot more complex compared to simpler barplot. Instead, the heatmap could be in this format.

![image](https://github.com/user-attachments/assets/19dda2f0-3d28-4f14-87ae-5d59d6a3cfc1)


Current heatmap

![image](https://github.com/user-attachments/assets/2f4542e3-eae5-4185-bd4c-90362ad53749)

Barplot with facets

![image](https://github.com/user-attachments/assets/0acbd54e-eed5-4a78-aab0-a7f278696617)

TuomasBorman commented:
After thinking this, we could have 2 layouts: "barplot" and "heatmap". The we could have hidden parameter "add.tree" that decides whether to add tree to heatmap (the "tree" layout is also one kind of heatmap)

The implementation could be then simplified a lot without losing features. This helps in the maintenance.

Commit:
Review changes

Commit:
Fix conflicts

Commit:
Changed heatmap layout & removed tree layout for add.tree parameter

Commit:
Workflow build fixes

ElySeraidarian commented:
Changes have been made, heatmap has been changed and parameter add.tree instead of tree layout

Commit:
Fix build

TuomasBorman committed:
Simplify

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,386 @@
+    if (is.null(rank)) {
+        df <- data.frame(Class = rownames(x))
+        legend_name <- "Class"
+    } else {
```
I think this plotter should take preprocessed data as input. The processing should happen in different internal function

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,386 @@
+  # Apply the function to each component and return the list
+  res <- lapply(seq_len(ncomponents), .process_component, df = df, n = n)
+  # Combine to single data.frame
+  res <- do.call(rbind, res)
```
I simplified this, check that it still works. 

instead of `1:ncomponents`, use `seq_len(ncomponents)`

TuomasBorman commented on R/plotLoadings.R:
```diff
@@ -0,0 +1,386 @@
+        
+        # Plot features with heatmap layout
+        if (layout == "heatmap") {
+            p <- ComplexHeatmap::Heatmap(x, heatmap_legend_param = list(title = "Value"))
```
This looks good, but the problem is that all other methods output ggplot objects. --> create a function that gets data from

.get_loadings_plot_data(x, n, ncomponents)

as input and returns heatmap created with ggplot2

antagomir committed:
Merge branch 'devel' into plotFeatureLoadings

Commit:
Review changes

Commit:
theme_minimal

Commit:
Fixed indentations

TuomasBorman commented on DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Package: miaViz
 Title: Microbiome Analysis Plotting and Visualization
-Version: 1.13.9
+Version: 1.14.0
```
Version should be 1.13.10

1.xx.yy

- xx denotes the Bioc release/devel version (updates every 6 months)
- yy rolling number

I will check this PR tomorrow at latest

TuomasBorman committed:
up

TuomasBorman commented:
I simplified the code a lot. Check that it still works as expected. Also run BiocCheck::BiocCheck() and fix issues related to this added R file (especially, do not exceed 80 characters in line width). Also use inheritParam tag in the documentation.

TuomasBorman commented:
Thing to discuss: should `rank` be `row.var`? --> variable defining a column from rowData. That is what it is doing currently

ElySeraidarian commented:
> Thing to discuss: should `rank` be `row.var`? --> variable defining a column from rowData. That is what it is doing currently

Yes I believe this would make more sense

antagomir commented:
Seems good.

ElySeraidarian commented:
But we could also still create a parameter rank that agglomerates tree as in the other functions

TuomasBorman commented:
> But we could also still create a parameter rank that agglomerates tree as in the other functions

If we agglomerate the data, we lose the match between loadings and rows. Loadings are for features of un-agglomerated data --> that is my first impression, check if that is correct

ElySeraidarian commented:
> > But we could also still create a parameter rank that agglomerates tree as in the other functions
> 
> If we agglomerate the data, we lose the match between loadings and rows. Loadings are for features of un-agglomerated data --> that is my first impression, check if that is correct

Yes that's right, agglomeration has to be done before performing the reduction method

Commit:
Fixes

ElySeraidarian committed:
Merge branch 'devel' into plotFeatureLoadings

TuomasBorman committed:
up

TuomasBorman commented:
1) I removed authors. See the issue: https://github.com/microbiome/mia/issues/628
2) You should perform first agglomeration and then transform the data (fixed)
3) I added a check that the feature names match in tree heatmap

TuomasBorman commented:
Looks very nice. Are you interested in updating OMA: https://microbiome.github.io/OMA/docs/devel/pages/beta_diversity.html#sec-dbrda-workflow

If I remember correctly, there were some error?

TuomasBorman committed:
up

ElySeraidarian commented:
> Looks very nice. Are you interested in updating OMA: https://microbiome.github.io/OMA/docs/devel/pages/beta_diversity.html#sec-dbrda-workflow
> 
> If I remember correctly, there were some error?

I will update using this function, I found same results when plotting. Also this example shows how to retrieve the loadings (for now we can plot only by getting matrix and then giving the matrix to plot).
